### PR TITLE
cucumber-expressions: Fix suggestions with in-word parameters

### PR DESCRIPTION
**Pull request by vincent-psarga**
## Summary

Do not suggest in-word parameters for cucumber expressions. For example, we should not suggest:

```
 I use gem I{int}n
```

for the sentence:

```
I use gem i18n
```

See issue #657

## Details

I think we'll need discussions about this one as some examples used in the test show the tool working the other way around.
For example in Ruby, we have this test that clearly shows that we may want to suggest in-word parameters:

```ruby
        expression = @generator.generate_expressions("I reach Stage4: 1st flight-1st hotel")[0]
        # While you would expect this to be `I reach Stage{int}: {optional-flight}-{optional-hotel}`
        # the `-1` causes {int} to match just before {optional-hotel}.
        expect(expression.source).to eq("I reach Stage{int}: {optional-flight}{int}st hotel")
```

What I did so far (for compatibility):
 - when creating a custom parameter type, we have an option "force_full_word" which is true by default. So in-word parameters won't be suggested
 - default types (int, float etc) have the "force_full_word" set to false (so the tests don't fail and we have the same behavior).

Before applying the same changes to Go, Java etc, I'd like to be sure this approach is correct (or should I remove this flag and never propose in-word parameters ? )

- [x] Bug fix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).

## Checklist:

- [x] The change has been ported to Java.
- [x] The change has been ported to Ruby.
- [x] The change has been ported to JavaScript.
- [x] The change has been ported to Go.
- [x] I've added tests for my code.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: cucumber-expressions/ruby/spec/cucumber/cucumber_expressions/cucumber_expression_generator_spec.rb [hunk 1]
```diff
@@ -162,6 +164,67 @@ class Currency
...
+            expect(@generator.generate_expression("When I watch the muppet show").source).not_to eq("When I watch the m{direction}pet show")
+            expect(@generator.generate_expression("When I watch the muppet show").source).to eq("When I watch the muppet show")
+          end
+
+          it "at the end of a word" do
+            expect(@generator.generate_expression("When I create a group").source).not_to eq("When I create a gro{direction}")
+            expect(@generator.generate_expression("When I create a group").source).to eq("When I create a group")
+          end
+
+          it "a full word" do
```
Comment (luke-hill) [inline cucumber-expressions/ruby/spec/cucumber/cucumber_expressions/cucumber_expression_generator_spec.rb]: this feels a separate context but not of big importance?

Comment (vincent-psarga) [reply cucumber-expressions/ruby/spec/cucumber/cucumber_expressions/cucumber_expression_generator_spec.rb]: I wanted to be sure not breaking anything there, but yes they are not that important checks indeed.

Maybe only keep the last one (which ensure that full words at the beginning and end of the sentence are matched)

File: cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/parameter_type.rb [hunk 2]
```diff
@@ -6,7 +6,7 @@ class ParameterType
       ILLEGAL_PARAMETER_NAME_PATTERN = /([\[\]()$.|?*+])/
       UNESCAPE_PATTERN = /(\\([\[$.|?*+\]]))/
 
-      attr_reader :name, :type, :regexps
+      attr_reader :name, :type, :regexps, :force_full_word
```
Comment (mpkorstanje) [inline cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/parameter_type.rb]: I hope I'm commenting on the right line.

I don't think this should be a flag. I would consider the old behavior to be just wrong so no need to enable it.

File: cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/parameter_type_matcher.rb [hunk 3]
```diff
@@ -3,26 +3,30 @@ module CucumberExpressions
...
+        @parameter_type, @regexp, @text, @force_full_word = parameter_type, regexp, text, force_full_word
         @match = @regexp.match(@text, match_position)
       end
 
       def advance_to(new_match_position)
         (new_match_position...@text.length).each {|advancedPos|
-          matcher = self.class.new(parameter_type, @regexp, @text, advancedPos)
-          if matcher.find
+          matcher = self.class.new(parameter_type, @regexp, @text, advancedPos, @force_full_word)
+          if matcher.find && matcher.full_word?
```
Comment (mpkorstanje) [inline cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/parameter_type_matcher.rb]: I would have expected this flag to be used in the Cucumber Expression Generator. What is the effect of using it here?

Comment (mpkorstanje) [reply cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/parameter_type_matcher.rb]: I get the impression this actually changes how the matching works.

Comment (mpkorstanje) [reply cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/parameter_type_matcher.rb]: Owh wait. I get it now. It's matcher. Not the parameter type.

**Comment by mpkorstanje** [comment]
>  For example in Ruby, we have this test that clearly shows that we may want to suggest in-word parameters:

I would consider a `-` a valid word boundary so it'd match like this:

```
        expect(expression.source).to eq("I reach Stage4: {optional-flight}-{optional-hotel}")
```

**Comment by mpkorstanje** [comment]
I think the following characters are valid boundaries ` `, `-`, `,`, `!`, `;`, `:`. They're all punctuation that normally follows a word.

**Comment by mpkorstanje** [comment]
I think that with possibly a few exceptions every punctuation mark should count as a boundary. There are some Unicode classes we could use for that.

https://en.wikipedia.org/wiki/Punctuation

**Comment by vincent-psarga** [comment]
> I think that with possibly a few exceptions every punctuation mark should count as a boundary. There are some Unicode classes we could use for that.
> 
> https://en.wikipedia.org/wiki/Punctuation

Definitely, that's something I didn't think about thanks. That should be taken into account now.

File: cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/parameter_type.rb (hunk 2 above)
Comment (vincent-psarga) [reply cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/parameter_type.rb]: I'm wondering about (which is the reason why so far only ruby is implemented).

I think it could be helpful for advanced users who may want to finely tweak the way suggestions are done.
But in the other hand those people would be mature enough to know how to tweak the step definitions and not blindly copy-paste what has been suggested by the tool.

Well, I'm a bit lost here :D

File: cucumber-expressions/ruby/spec/cucumber/cucumber_expressions/cucumber_expression_generator_spec.rb (hunk 1 above)
Comment (luke-hill) [reply cucumber-expressions/ruby/spec/cucumber/cucumber_expressions/cucumber_expression_generator_spec.rb]: I think I may have been mis-interpreted.

The checks are fine. But I would wrap them as thus

```
context 'when set to true, does suggest parameter when a match is'
```

File: cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/parameter_type.rb (hunk 2 above)
Comment (mpkorstanje) [reply cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/parameter_type.rb]: In Java this would mean yet another constructor variation. This makes the API more difficult to use because (when your IDE isn't perfectly) configured you'd have to guess between `....boolean, boolean)` and `...boolean, boolean, boolean)`. So I'd prefer not to have an extra one. 

Additionally the cucumber expression generator is already prone to producing combinatorial explosions. So anything that prunes the suggestions to include only the most likely ones is good.

File: cucumber-expressions/ruby/spec/cucumber/cucumber_expressions/cucumber_expression_generator_spec.rb (hunk 1 above)
Comment (vincent-psarga) [reply cucumber-expressions/ruby/spec/cucumber/cucumber_expressions/cucumber_expression_generator_spec.rb]: ahh ok I got it now sorry. Yes I can do that :)

File: cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/parameter_type_matcher.rb [hunk 4]
```diff
@@ -47,12 +47,12 @@ def <=>(other)
 
       def space_before_match_or_sentence_start
         match_begin = @match.begin(0)
-        match_begin == 0 || @text[match_begin - 1].match(/\s/)
+        match_begin == 0 || @text[match_begin - 1].match(/\s|[[:punct:]]/)
```
Comment (mpkorstanje) [inline cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/parameter_type_matcher.rb]: As far as I know `\s` only matches `[ \t\r\n\]`. However there are other spacing marks in use. In Hindi, Arabic and Pashto for example. (I don't have my reference examples around at the moment). So I think the unicode class `/\p{Zs}/ ` might be more appropriate here.

I believe `|[[:punct:]]` is a ruby specific implementation. To keep the implementation consistent between implementations I think it might be better to use Unicode character classes. All languages are able to understand these. For punctuation that would be `/\p{P}/`.

**Comment by aslakhellesoy** [comment]
> I remember when I raised that original issue. The problem was originally an infinite loop which aslak did fix up a while back.
> 
> I think the usage was because often when you want an alternation in normal text you would right `1 or 2` as `1/2`, but this isn't possible using cuke-exp, because `/` is alternation. So I opted for `1-2` but this then got parsed as a very complex combination of integers.

@luke-hill I think the issue referring to is this: https://github.com/cucumber/cucumber/issues/307

File: cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/parameter_type.rb (hunk 2 above)
Comment (aslakhellesoy) [reply cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/parameter_type.rb]: I agree with @mpkorstanje - we should not make this a configurable flag. YAGNI, KISS etc.

File: cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/parameter_type_matcher.rb (hunk 4 above)
Comment (aslakhellesoy) [reply cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/parameter_type_matcher.rb]: For the Java implementation I went for `[\s\p{P}]` to determine if a character is whitespace or punctuation. I didn't look into Hindi/Arabic/Pashto spacing. I think we can defer supporting that until it becomes an issue. I prefer simplicity over completeness for this functionality - it's not execution, it's just snippet suggestions.

File: cucumber-expressions/java/src/test/java/io/cucumber/cucumberexpressions/CucumberExpressionGeneratorTest.java [hunk 5]
```diff
@@ -234,43 +273,6 @@ public void exposes_transforms_in_generated_expression() {
...
-                true,
-                false
-        );
-
-        parameterTypeRegistry.defineParameterType(optionalFlight);
-        parameterTypeRegistry.defineParameterType(optionalHotel);
-        // While you would expect this to be `I reach Stage{int}: {optional-flight}-{optional-hotel}` the `-1` causes
-        // {int} to match just before {optional-hotel}.
-        List<GeneratedExpression> generatedExpressions = generator.generateExpressions("I reach Stage4: 1st flight-1st hotel");
-        assertEquals("I reach Stage{int}: {optional-flight}{int}st hotel", generatedExpressions.get(0).getSource());
```
Comment (mpkorstanje) [inline cucumber-expressions/java/src/test/java/io/cucumber/cucumberexpressions/CucumberExpressionGeneratorTest.java]: I'm would not expect this to be deleted. This should now match `"I reach Stage{int}: {optional-flight}{optional-hotel}" because the `-` is punctuation.

File: cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/parameter_type_matcher.rb (hunk 4 above)
Comment (mpkorstanje) [reply cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/parameter_type_matcher.rb]: Good enough for now I guess. I'll add the extra character classes + tests when I get the time. It doesn't add much complexity. It's all in the Unicode standard.

File: cucumber-expressions/java/src/test/java/io/cucumber/cucumberexpressions/CucumberExpressionGeneratorTest.java (hunk 5 above)
Comment (aslakhellesoy) [reply cucumber-expressions/java/src/test/java/io/cucumber/cucumberexpressions/CucumberExpressionGeneratorTest.java]: You're right @mpkorstanje - deleting it was perhaps a bit sloppy of me. The thing is, I couldn't make the test verify the behaviour it was initially written to test, namely that the optional group had a 0-length match. We could keep it, tweak it and make it pass as you suggested, but it would be testing behaviour we're already testing I think...

File: cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/parameter_type_matcher.rb (hunk 4 above)
Comment (vincent-psarga) [reply cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/parameter_type_matcher.rb]: Thanks for pointing that out, the Ruby implementation now uses \p{P} instead of [[:punct]]

File: cucumber-expressions/java/src/test/java/io/cucumber/cucumberexpressions/CucumberExpressionGeneratorTest.java (hunk 5 above)
Comment (vincent-psarga) [reply cucumber-expressions/java/src/test/java/io/cucumber/cucumberexpressions/CucumberExpressionGeneratorTest.java]: ON the Ruby implementation, I've had to update this test so it still reflects the expected behaviour (matching the int before the optional capture group).
The sentence does not make that much sense now, hopefully the comment below still explain correctly why we don't match (optional-hotel)

File: cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/parameter_type.rb (hunk 2 above)
Comment (vincent-psarga) [reply cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/parameter_type.rb]: that's been removed.

File: cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/parameter_type_matcher.rb [hunk 6]
```diff
@@ -4,14 +4,14 @@ class ParameterTypeMatcher
       attr_reader :parameter_type
 
       def initialize(parameter_type, regexp, text, match_position=0)
-        @parameter_type, @regexp, @text = parameter_type, regexp, text
+        @parameter_type, @regexp, @text, @force_full_word = parameter_type, regexp, text
```
Comment (luke-hill) [inline cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/parameter_type_matcher.rb]: Think this can be removed

File: cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/parameter_type_matcher.rb [hunk 7]
```diff
@@ -38,6 +42,18 @@ def <=>(other)
         return length_comparison if length_comparison != 0
         0
       end
+
+      private
+
+      def space_before_match_or_sentence_start
```
Comment (luke-hill) [inline cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/parameter_type_matcher.rb]: Think this can be a predicate - so `space_before_match_or_sentence_start?`

File: cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/parameter_type_matcher.rb [hunk 8]
```diff
@@ -38,6 +42,18 @@ def <=>(other)
...
+        match_begin = @match.begin(0)
+        match_begin == 0 || @text[match_begin - 1].match(/\s|\p{P}/)
+      end
+
+      def space_after_match_or_sentence_end
```
Comment (luke-hill) [inline cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/parameter_type_matcher.rb]: This should probably end in `?` too.

File: cucumber-expressions/ruby/spec/cucumber/cucumber_expressions/cucumber_expression_generator_spec.rb [hunk 9]
```diff
@@ -162,6 +162,58 @@ class Currency
         expect(expressions[1].source).to eq("{zero-or-more} {zero-or-more} {zero-or-more}")
       end
 
+      context "does not suggest parameter when match is" do
+        let!(:direction_parameter_type) {
```
Comment (luke-hill) [inline cucumber-expressions/ruby/spec/cucumber/cucumber_expressions/cucumber_expression_generator_spec.rb]: not of huge importance, but do these need to be let bangs? It'll cause some issues for local runs potentially. It'll also annoy rubocop.

These aren't blockers. And if needs to remain, I'm :cool: with that. But if we can change them just to `let` that would be better

Comment (vincent-psarga) [reply cucumber-expressions/ruby/spec/cucumber/cucumber_expressions/cucumber_expression_generator_spec.rb]: It's done :)

For the let! part, it was useful in the previous version to be able to use let(:force_full_word) and override it in sub-context but that does not make sense anymore indeed :)

**Comment by mpkorstanje** [comment]
I wouldn't mind porting this one over to java and I'll review the go and javascript implementations afterwards.

**Comment by aslakhellesoy** [comment]
> I wouldn't mind porting this one over to java and I'll review the go and javascript implementations afterwards.

@mpkorstanje - I've already ported this to Java (in this PR)

File: cucumber-expressions/java/src/main/java/io/cucumber/cucumberexpressions/ParameterTypeRegistry.java [hunk 10]
```diff
@@ -19,7 +19,7 @@
     // Pattern.compile(...).pattern() is not necessary, but it helps us take advantage of the IntelliJ's regexp validation,
     // which detects unneeded escapes.
     private static final List<String> INTEGER_REGEXPS = asList(Pattern.compile("-?\\d+").pattern(), Pattern.compile("\\d+").pattern());
-    private static final List<String> FLOAT_REGEXPS = singletonList(Pattern.compile("-?\\d*(?:[.,]\\d+)?").pattern());
+    private static final List<String> FLOAT_REGEXPS = singletonList(Pattern.compile("[+-]?(?:[0-9]*[.,])?[0-9]+").pattern());
```
Comment (mpkorstanje) [inline cucumber-expressions/java/src/main/java/io/cucumber/cucumberexpressions/ParameterTypeRegistry.java]: @aslakhellesoy why did you change this? And why only for Java?

File: cucumber-expressions/java/src/main/java/io/cucumber/cucumberexpressions/ParameterTypeMatcher.java [hunk 11]
```diff
@@ -1,30 +1,46 @@
...
+                    char before = text.charAt(matcher.start() - 1);
+                    if (!isWhitespaceOrPunctuation(before)) {
+                        return false;
+                    }
+                }
+
+                if (matcher.end() < text.length()) {
+                    char after = text.charAt(matcher.end());
+                    return isWhitespaceOrPunctuation(after);
+                }
```
Comment (mpkorstanje) [inline cucumber-expressions/java/src/main/java/io/cucumber/cucumberexpressions/ParameterTypeMatcher.java]: I find this section really hard to follow. Perhaps we extract a method here that makes it clear we should have a match and that the match should be a full word.

Comment (mpkorstanje) [reply cucumber-expressions/java/src/main/java/io/cucumber/cucumberexpressions/ParameterTypeMatcher.java]: So something like 
```java
             if (!group().isEmpty() && matchIsFullWord()) { 
```

**Comment by mpkorstanje** [comment]
Right. I even commented on it.

File: cucumber-expressions/java/src/main/java/io/cucumber/cucumberexpressions/ParameterTypeRegistry.java (hunk 10 above)
Comment (aslakhellesoy) [reply cucumber-expressions/java/src/main/java/io/cucumber/cucumberexpressions/ParameterTypeRegistry.java]: When I fiddled with the test I removed I discovered that an undefined step def with `-` would suggest {float}. That's why I changed it. In retrospect I realise that should be done in a different PR, so I'll remove it from this one.

**Comment by mpkorstanje** [comment]
I'd say this is ready to go. I'll merge it at the end of next week and bundle it up with the other cucumber expression mrs for a release.  Please do feel free to merge earlier though.
